numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -2236,14 +2240,25 @@ VOID_copyswapn (char *dst, npy_intp dstride, char *src, npy_intp sstride,
         ((PyArrayObject_fields *)arr)->descr = descr;
         return;
     }
-    if (swap && PyArray_DESCR(arr)->subarray != NULL) {
+    if (descr->subarray == NULL) {
+        /*
+         * Cannot have objects or incontiguity without fields or subarrays.
+         * So copying the whole elsize is correct. (Must be plain VOID)
+         */
+        _basic_copyn(dst, dstride, src, sstride, n, descr->elsize);
+        return;
+    }
+    if (swap ||
+            /* Datatype may require the copyswap function (see also else): */
+            PyDataType_HASFIELDS(descr->subarray->base) ||
```

Comment:
Why are we checking for fields again, when that path was already hit above?

Reply:
This is on the subarray, not the original.

Reply:
A, maybe should include subarray in the comment...

Reply:
Can you extract a `npy_bool can_optimize_subarray = ...` variable here, since this is really just trying to be an optimization? I think that should look like:
```
npy_bool can_optimize_subarray = !swap && !is_object && !has_fields && is_builtin;
```